src/main/scala/io/iohk/ethereum/jsonrpc/QAService.scala

```diff
@@ -54,6 +57,46 @@ class QAService(
     getTransactionsFromPool.map { resp =>
       Right(GetPendingTransactionsResponse(resp.pendingTransactions))
     }
+
+  def generateCheckpoint(
+      req: GenerateCheckpointRequest
+  ): ServiceResponse[GenerateCheckpointResponse] = {
+    Future {
+      req.blockHash match {
+        case Some(hash) =>
+          val checkpoint = generateCheckpoint(hash, req.privateKeys)
+          syncController ! NewCheckpoint(hash, checkpoint.signatures)
+          Right(GenerateCheckpointResponse(checkpoint))
+        case None =>
+          val hash = blockchain.getBestBlock().hash
+          val checkpoint = generateCheckpoint(hash, req.privateKeys)
+          syncController ! NewCheckpoint(hash, checkpoint.signatures)
+          Right(GenerateCheckpointResponse(checkpoint))
+      }
+    } recover { case t: Throwable =>
+      log.info("Unable to generate requested checkpoint", t)
+      Left(JsonRpcErrors.InternalError)
+    }
+  }
+
+  private def generateCheckpoint(blockHash: ByteString, privateKeys: Seq[ByteString]): Checkpoint = {
+    val keys = privateKeys.map { key =>
+      crypto.keyPairFromPrvKey(key.toArray)
+    }
+    val signatures = keys.map(ECDSASignature.sign(blockHash.toArray, _, None))
```

Comment:
Could you add TODO about https://github.com/input-output-hk/mantis/pull/738 to remember to add sorting here too?

Reply:
See [this comment](https://github.com/input-output-hk/mantis/pull/738#issuecomment-709387397)